src/PowerShellEditorServices.Host/CodeLens/CodeLensFeature.cs

```diff
@@ -67,42 +68,82 @@ internal class CodeLensFeature :
             return codeLenses;
         }
 
+        /// <summary>
+        /// The editor session context to get workspace and language server data from.
+        /// </summary>
+        private readonly EditorSession _editorSession;
+
+        /// <summary>
+        /// The json serializer instance for CodeLens object translation.
+        /// </summary>
+        private readonly JsonSerializer _jsonSerializer;
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="editorSession"></param>
+        /// <param name="jsonSerializer"></param>
+        /// <param name="logger"></param>
+        private CodeLensFeature(
+            EditorSession editorSession,
+            JsonSerializer jsonSerializer,
+            ILogger logger)
+                : base(logger)
+        {
+            _editorSession = editorSession;
+            _jsonSerializer = jsonSerializer;
+        }
+
+        /// <summary>
+        /// Get all the CodeLenses for a given script file.
+        /// </summary>
+        /// <param name="scriptFile">The PowerShell script file to get CodeLenses for.</param>
+        /// <returns>All generated CodeLenses for the given script file.</returns>
         public CodeLens[] ProvideCodeLenses(ScriptFile scriptFile)
         {
-            return
-                this.InvokeProviders(p => p.ProvideCodeLenses(scriptFile))
-                    .SelectMany(r => r)
-                    .ToArray();
+            var acc = new List<CodeLens>();
+            foreach (CodeLens[] providerResult in InvokeProviders(provider => provider.ProvideCodeLenses(scriptFile)))
+            {
+                acc.AddRange(providerResult);
+            }
+
+            return acc.ToArray();
```

Comment:
I would think allocating a list and then an array would end up with more allocations.  I don't know enough about how `SelectMany` is implemented to say that with any certainty, but maybe a quick benchmark test would be a good idea.

Speaking of, benchmarks is a good bullet point for #720

Reply:
So I wrote a nasty little benchmark:
```csharp
    class Program
    {
        private static Random s_random = new Random();

        static void Main(string[] args)
        {
            var tests = new Dictionary<string, Action>
            {
                { "Select Many Concatenation", ConcatWithSelectMany },
                { "List Build Concatentation", ConcatWithList },
            };

            RunBenchmark(tests);
        }

        private static void ConcatWithSelectMany()
        {
            GenerateMultiData(d => d.ToTransformedDummyData()).SelectMany(d => d).ToArray();
        }

        private static void ConcatWithList()
        {
            var acc = new List<TransformedDummyData>();
            foreach (TransformedDummyData[] dataArr in GenerateMultiData(d => d.ToTransformedDummyData()))
            {
                acc.AddRange(dataArr);
            }
            acc.ToArray();
        } 

        private static IEnumerable<TransformedDummyData[]> GenerateMultiData(Func<DummyData, TransformedDummyData> process, int quantity = 5)
        {
            for (int i = 0; i < quantity; i++)
            {
                yield return GenerateData(s_random.Next(5, 15)).Select(process).ToArray();
            }
        }

        private static IEnumerable<DummyData> GenerateData(int quantity = 10)
        {
            for (int i = 0; i < quantity; i++)
            {
                yield return new DummyData(s_random.Next(), new object(), Guid.NewGuid().ToString());
            }
        }

        private static IntervalData Time(string testName, Action action, int repeats = 1000, int subInterval = 100)
        {
            var times = new List<double>(repeats);
            for (int i = 0; i < repeats; i++)
            {
                var watch = Stopwatch.StartNew();
                for (int j = 0; j < subInterval; j++)
                {
                    action();
                }
                watch.Stop();
                times.Add((double)watch.ElapsedMilliseconds / subInterval);
            }

            double maxTime = times.Max();
            double minTime = times.Min();
            double meanTime = times.Average();
            double[] sortedTimes = times.OrderBy(t => t).ToArray();

            return new IntervalData
            {
                TestName = testName,
                MaxTime = maxTime,
                MaxTimeIndex = times.IndexOf(maxTime),
                MinTime = minTime,
                MinTimeIndex = times.IndexOf(minTime),
                TotalTime = times.Sum(),
                MeanTime = meanTime,
                StdDev = Math.Sqrt(times.Sum(t => Math.Pow(t - meanTime, 2)) / (times.Count - 1)),
                MedianTime = sortedTimes.Length % 2 == 0 ? sortedTimes[sortedTimes.Length / 2] : sortedTimes[sortedTimes.Length / 2 + 1]
            };
        }

        private static void RunBenchmark(IDictionary<string, Action> tests)
        {
            foreach (KeyValuePair<string, Action> test in tests)
            {
                IntervalData result = Time(test.Key, test.Value);
                Console.WriteLine($"Test {result.TestName}:");
                Console.WriteLine($"\tTotal Time:\t\t\t{result.TotalTime}");
                Console.WriteLine($"\tMean Time:\t\t\t{result.MeanTime}");
                Console.WriteLine($"\tMedian Time:\t\t\t{result.MedianTime}");
                Console.WriteLine($"\tStd Dev:\t\t\t{result.StdDev}");
                Console.WriteLine($"\tMax Time:\t\t\t{result.MaxTime}");
                Console.WriteLine($"\tMax Time Index:\t\t\t{result.MaxTimeIndex}");
                Console.WriteLine($"\tMin Time:\t\t\t{result.MinTime}");
                Console.WriteLine($"\tMin Time Index:\t\t\t{result.MinTimeIndex}");
                Console.WriteLine();
            }
        }
    }

    public class DummyData
    {
        public DummyData(int x, object y, string z)
        {
            X = x;
            Y = y;
            Z = z;
        }

        public int X { get; }

        public object Y { get; }

        public string Z { get; }
    }

    public static class DummyDataExtensions
    {
        public static TransformedDummyData ToTransformedDummyData(this DummyData data)
        {
            return new TransformedDummyData(data.X+1, data.Z?.ToUpper());
        }
    }

    public class TransformedDummyData
    {
        public TransformedDummyData(int x, string z)
        {
            X = x;
            Z = z;
        }

        public int X { get; }

        public string Z { get; }
    }

    public class IntervalData
    {
        public string TestName;

        public double MaxTime;

        public int MaxTimeIndex;

        public double MinTime;

        public int MinTimeIndex;

        public double TotalTime;

        public double MeanTime;

        public double MedianTime;

        public double StdDev;
    }
```

With the following results:
```
Test Select Many Concatenation:
	Total Time:			60.7800000000007
	Mean Time:			0.0607800000000007
	Median Time:			0.06
	Std Dev:			0.00494125652198061
	Max Time:			0.13
	Max Time Index:			0
	Min Time:			0.06
	Min Time Index:			3

Test List Build Concatentation:
	Total Time:			60.4900000000008
	Mean Time:			0.0604900000000008
	Median Time:			0.06
	Std Dev:			0.00472040063788888
	Max Time:			0.13
	Max Time Index:			431
	Min Time:			0.06
	Min Time Index:			1


```

Don't know how representative my benchmark really is, but given that result, I will revert to `SelectMany` since it's simpler and may end up allocating less in the long run.

Reply:
Awesome!

I converted your benchmark to use [BenchmarkDotNet](https://gist.github.com/SeeminglyScience/69483641e45f90cf31d2ce948428b288).

Here's the results I got in CoreCLR/Windows

Method |     Mean |     Error |    StdDev | Scaled | ScaledSD |  Gen 0 | Allocated |
--------------------- |---------:|----------:|----------:|-------:|---------:|-------:|----------:|
 ConcatWithSelectMany | 36.28 us | 0.7209 us | 0.8302 us |   1.00 |     0.00 | 5.4321 |  16.82 KB |
 ConcatWithList | 38.01 us | 0.9687 us | 2.8562 us |   1.05 |     0.08 | 5.6763 |   17.6 KB |